step-registry: rework "Parameters" section

bbguimaraes opened the pull request:
Reviewing will probably be easier by simply ignoring the red `-` lines or, even
better, just looking at the new `#parameters` and `#hierarchical-propagation`
sections.

https://deploy-preview-141--zealous-shirley-0f56b3.netlify.app/docs/architecture/step-registry#parameters
https://deploy-preview-141--zealous-shirley-0f56b3.netlify.app/docs/architecture/step-registry#hierarchical-propagation

bbguimaraes commented:
https://issues.redhat.com/browse/DPTP-2146

bbguimaraes commented:
https://github.com/openshift/ci-docs/pull/142 fixes the (unrelated) out-of-date URLs.

stevekuznetsov commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -409,86 +427,79 @@ ref:
+{{< alert title="Warning" color="warning" >}}
+Parameters must be declared in the `env` section of every step that requires
+them, as shown here. Simply setting the parameter in a parent component is not
+sufficient.
```
Add the why - it's due to the fact that steps need to declare they consume it, or we won't give it to them

bbguimaraes replied:
I was trying to treat this as the "I don't care" section and tried not to repeat the one in the other.

Oh well… duplicated the other warning.

stevekuznetsov commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -409,86 +427,79 @@ ref:
+Note that `env` is a top-level field in steps (and chains), alongside the `as`
+field, while it is placed in the `steps` field in tests (and workflows). The
+strict YAML validation used to parse these files will generate an error, but
+this is still a common source of confusion.
```
Maybe an example would help?

bbguimaraes replied:
I was hoping the two right above would do it.

stevekuznetsov commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -559,11 +570,141 @@ tests:
+One special rule applies to list and mapping fields that are specified both in a
+test and its workflow.  Instead of completely overriding the workflow value, as
+is the case for scalar values, the two sections are merged according to the
+following rules:
```
An example would help here, too

bbguimaraes replied:
Added.

stevekuznetsov reviewed (commented):
This is much clearer (at least to me!)

stevekuznetsov commented:
Seems reasonable, but perhaps the hierarchy examples could be improved, they are a bit of a wall of text right now. @petr-muller @emilvberglind  PTAL?

petr-muller commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -409,86 +427,81 @@ ref:
+    - ref: openshift-e2e-test
+    env:
+      TEST_SUITE: openshift/conformance/parallel
+{{< / highlight >}}
```
It is unclear what this snippet purpose is. It immediately follows the "Failing to do so will result in a validation error" sentence so I expected it to show the "wrong" configuration, but it is actually correct. So I think the snippet needs a lead paragraph that says what it shows.

petr-muller commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -409,86 +427,81 @@ ref:
+Note that `env` is a top-level field in steps (and chains), alongside the `as`
+field, while it is placed in the `steps` field in tests (and workflows). The
+strict YAML validation used to parse these files will generate an error, but
+this is still a common source of confusion.
```
The two alert boxes without any accompanying text look awkward when rendered. The alert boxes are usually used to stress out some important aspect of the base text or clarify a potential pitfall or something like that, but here there's no base text in the fuirst place.

petr-muller commented on content/en/docs/architecture/step-registry.md:
```diff
@@ -559,11 +572,176 @@ tests:
+
+### Examples
+
+A step declaring a parameter:
```
I think giving the individual examples short names and using L4 headers (####) would help to break this section down visually (and individual examples would be linkable which may be useful)

petr-muller reviewed (commented):
It's definitely better than it was before! I think the `Parameters` section could be improved by adding two subsections that reflect the two types of users that need to read this doc. One type of user writes step registry content, and needs to understand how to specify what parameters their steps/chains can receive. The second type of user is the test writer, and needs to understand how to set values for the steps/chains.

The section right now kinda serves both by intermixing them, but I think we would win if we had separate "Read this if you write a step" and "Read this if you write a test" sections.

bbguimaraes commented:
Intermediary update:

- f7410c6: added a "common mistakes" section and moved warnings to it, leaving sporadic references in the original text.

  The initial reasoning for the warnings and their placement was that experience has shown they just will not be read if they are not close to the definition/example and prominent.  Hopefully the section name will be sufficiently attractive.

- 73b4a16: separated the main section as requested.  The first part is mostly unchanged.  I added links to relevant pages that show available registry components and their parameters to the second.

I'll work on the other section next. Many of the examples in it feel redundant now with the already somewhat-duplicated examples in the main section, perhaps it will be more focused with some pruning and adjustments.

bbguimaraes commented:
@petr-muller, @stevekuznetsov: PTAL again.

petr-muller reviewed (approved):
Nice job!
